File: packages/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/ReanimatedModule.java
package com.swmansion.reanimated;

import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;

@ReactModule(name = ReanimatedModule.NAME)
public class ReanimatedModule extends NativeReanimatedModuleSpec implements LifecycleEventListener {
  private NodesManager mNodesManager;
  private boolean mTurboModuleInstalled = false;

  public ReanimatedModule(ReactApplicationContext reactContext) {
    super(reactContext);
    reactContext.assertOnJSQueueThread();
  }

  @Override
  public void initialize() {
    ReactApplicationContext reactContext = getReactApplicationContext();
    reactContext.assertOnJSQueueThread();
    reactContext.addLifecycleEventListener(this);
  }

  @Override
  public void onHostPause() {
    if (mNodesManager != null) {
      mNodesManager.onHostPause();
    }
  }

  @Override
  public void onHostResume() {
    if (mNodesManager != null) {
      mNodesManager.onHostResume();
    }
  }

  @Override
  public void onHostDestroy() {
    // do nothing
  }

  public NodesManager getNodesManager() {
    // This method is called from react-native-gesture-handler.
    return mNodesManager;
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public boolean installTurboModule() {
    ReactApplicationContext reactContext = getReactApplicationContext();
    reactContext.assertOnJSQueueThread();
    mNodesManager = new NodesManager(reactContext);
    mNodesManager.getNativeProxy().installJSIBindings();
    return true;
  }

  @ReactMethod
  public void addListener(String ignoredEventName) {
    // Keep: Required for RN built in Event Emitter Calls.
  }

  @ReactMethod
  public void removeListeners(Integer ignoredCount) {
    // Keep: Required for RN built in Event Emitter Calls.
  }

  @Override
  public void invalidate() {
    super.invalidate();
    if (mNodesManager != null) {
      mNodesManager.invalidate();
    }
    ReactApplicationContext reactContext = getReactApplicationContext();
    reactContext.removeLifecycleEventListener(this);
  }
}

File: packages/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/ReanimatedModule.kt
package com.swmansion.reanimated

import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.module.annotations.ReactModule

@ReactModule(name = NativeReanimatedModuleSpec.NAME)
class ReanimatedModule(
    reactContext: ReactApplicationContext,
) : NativeReanimatedModuleSpec(reactContext),
    LifecycleEventListener {
    private var mNodesManager: NodesManager? = null
    private var mTurboModuleInstalled = false

    init {
        reactContext.assertOnJSQueueThread()
    }

    override fun initialize() {
        val reactContext = reactApplicationContext
        reactContext.assertOnJSQueueThread()
        reactContext.addLifecycleEventListener(this)
    }

    override fun onHostPause() {
        mNodesManager?.onHostPause()
    }

    override fun onHostResume() {
        mNodesManager?.onHostResume()
    }

    override fun onHostDestroy() {
        // do nothing
    }

    // This property is accessed from react-native-gesture-handler.
    val nodesManager: NodesManager?
        get() = mNodesManager

    @ReactMethod(isBlockingSynchronousMethod = true)
    override fun installTurboModule(): Boolean {
        val reactContext = reactApplicationContext
        reactContext.assertOnJSQueueThread()
        mNodesManager = NodesManager(reactContext)
        mNodesManager!!.getNativeProxy()!!.installJSIBindings()
        return true
    }

    @ReactMethod
    fun addListener(ignoredEventName: String) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    fun removeListeners(ignoredCount: Int) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    override fun invalidate() {
        super.invalidate()
        mNodesManager?.invalidate()
        val reactContext = reactApplicationContext
        reactContext.removeLifecycleEventListener(this)
    }
}
